packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -154,33 +238,277 @@ export class BiconomySmartAccountV2 extends BaseSmartAccount {
   getDummyPaymasterData(): string {
     return '0x'
   }
 
-  // TODO
+  async signUserOp(userOp: Partial<UserOperation>): Promise<UserOperation> {
+    const requiredFields: UserOperationKey[] = [
+      'sender',
+      'nonce',
+      'initCode',
+      'callData',
+      'callGasLimit',
+      'verificationGasLimit',
+      'preVerificationGas',
+      'maxFeePerGas',
+      'maxPriorityFeePerGas',
+      'paymasterAndData'
+    ]
+    super.validateUserOp(userOp, requiredFields)
+    const userOpHash = await this.getUserOpHash(userOp)
+
+    // Note: we could use signUserOp from validationModule
+    const sig = await this.activeValidationModule.signMessage(arrayify(userOpHash))
+
+    userOp = {
+      ...userOp,
+      signature: sig
+    }
+    const signatureWithModuleAddress = ethers.utils.defaultAbiCoder.encode(
+      ['bytes', 'address'],
+      [userOp.signature, await this.activeValidationModule.getAddress()]
+    )
+    userOp.signature = signatureWithModuleAddress
+    return userOp as UserOperation
+  }
+
+  async buildUserOp(
+    transactions: Transaction[],
+    overrides?: Overrides,
+    skipBundlerGasEstimation?: boolean
+  ): Promise<Partial<UserOperation>> {
+    // Review: may not need at all
+    // this.isInitialized()
+
+    // TODO: validate to, value and data fields
+    // TODO: validate overrides if supplied
+    const to = transactions.map((element: Transaction) => element.to)
+    const data = transactions.map((element: Transaction) => element.data ?? '0x')
+    const value = transactions.map((element: Transaction) => element.value ?? BigNumber.from('0'))
+
+    let callData = ''
+    if (transactions.length === 1) {
+      callData = await this.encodeExecute(to[0], value[0], data[0])
+    } else {
+      callData = await this.encodeExecuteBatch(to, value, data)
+    }
+    let nonce = BigNumber.from(0)
+    try {
+      nonce = await this.getNonce()
+    } catch (error) {
+      // Not throwing this error as nonce would be 0 if this.getNonce() throw exception, which is expected flow for undeployed account
+    }
+
+    // let isDeployed = false
+
+    /*if (nonce.eq(0) && this.accountAddress) {
+      isDeployed = await this.isAccountDeployed(this.accountAddress)
+    }*/
+
+    let userOp: Partial<UserOperation> = {
+      sender: this.accountAddress,
+      nonce,
+      initCode: await this.getInitCode(),
+      callData: callData
+    }
+
+    // for this Smart Account current validation module dummy signature will be used to estimate gas
+    userOp.signature = this.getDummySignature()
+
+    userOp = await this.estimateUserOpGas(userOp, overrides, skipBundlerGasEstimation)
+    Logger.log('userOp after estimation ', userOp)
+
+    // Do not populate paymasterAndData as part of buildUserOp as it may not have all necessary details
+    userOp.paymasterAndData = '0x' // await this.getPaymasterAndData(userOp)
+
+    delete userOp.signature
+    return userOp
+  }
+
+  private validateUserOpAndRequest(
```

Comment:
Rename method to 
`validateUserOpAndRequest` to `validateUserOpAndPaymasterRequest`

Reply:
done